### PR TITLE
Fix loading data on different encoding

### LINKED ISSUE #65300: zip file serialization endianness detection is broken
## 🐛 Bug

When training and saving a model on a little endian platform, there does not seem to be any "host" endianness attribute carried into the saved model data.
When loaded on a big endian platform, the model parameters are incorrect; that is, the byteorder is not checked and corrected on model load.

## To Reproduce

1. Create a simple model as shown below, and set model weights. 
2. Save model state_dict
3. Load model on big endian platform.
4. Display weights

**Build and set model weights on LE system:**
```
lstm = torch.nn.LSTM(3, 3)  # Input dim is 3, output dim is 3
inputs = [torch.randn(1, 3) for _ in range(5)]  # make a sequence of length 5

inputs = torch.cat(inputs).view(len(inputs), 1, -1)
hidden = (torch.randn(1, 1, 3), torch.randn(1, 1, 3))  # clean out hidden state
```
```
print("Model weight weight_ih_l0:")
print(lstm.weight_ih_l0)
Model weight weight_ih_l0:
Parameter containing:
tensor([[ 0.3852,  0.1366,  0.0782],
        [ 0.1975, -0.4888,  0.3619],
        [ 0.0071,  0.3601,  0.4281],
        [-0.4235,  0.4426, -0.3384],
        [-0.2385,  0.4222,  0.1653],
        [-0.2102, -0.2984,  0.1414],
        [-0.0508,  0.5562, -0.1868],
        [ 0.2324, -0.3226, -0.5770],
        [-0.4104,  0.0577,  0.2712],
        [-0.4971,  0.2164,  0.1720],
        [-0.3956, -0.0690,  0.1853],
        [-0.2401, -0.1618, -0.5637]], requires_grad=True)
```
**Copy to big endian system, then load model:**
```
filepath = "lstm.pt"
lstm.load_state_dict(torch.load(filepath),strict=True) 
print(lstm.weight_ih_l0)
tensor([[ 1.4603e-25, -1.5686e-21,  1.2081e-19],
        [-2.6189e-12,  1.5163e+28, -3.3969e+12],
        [ 1.7073e-38,  6.6708e+25, -4.6704e-39],
        [-1.6697e-03, -2.6238e-19, -1.3629e+16],
        [-6.7119e-16, -1.0742e+04, -6.3514e-39],
        [ 5.6304e+19,  4.6716e+20, -1.9736e+27],
        [-1.8113e+29, -6.7717e+25, -9.6834e+26],
        [-1.0278e+37, -2.5245e-12, -9.8826e+10],
        [ 3.4006e-17,  1.4774e-19,  1.0029e-10],
        [-8.1701e+28,  6.6931e-05,  9.2278e-03],
        [-7.8279e+19,  5.4253e-05,  1.5379e+03],
        [-8.6688e+24,  5.1006e-06,  1.1713e-11]], requires_grad=True)
```
**Swap byte order on BE system to confirm:**
```

print("Byteswap state_dict")
for param_tensor in lstm.state_dict():
    print(param_tensor, "\t", lstm.state_dict()[param_tensor].size())
    print(lstm.state_dict()[param_tensor])
    updated = lstm.state_dict()[param_tensor].cpu().detach().numpy().byteswap()
    lstm.state_dict()[param_tensor][:] = torch.Tensor(updated)
    print(lstm.state_dict()[param_tensor])
```
_... trimmed results for readability_
```
tensor([[ 0.3852,  0.1366,  0.0782],
        [ 0.1975, -0.4888,  0.3619],
        [ 0.0071,  0.3601,  0.4281],
        [-0.4235,  0.4426, -0.3384],
        [-0.2385,  0.4222,  0.1653],
        [-0.2102, -0.2984,  0.1414],
        [-0.0508,  0.5562, -0.1868],
        [ 0.2324, -0.3226, -0.5770],
        [-0.4104,  0.0577,  0.2712],
        [-0.4971,  0.2164,  0.1720],
        [-0.3956, -0.0690,  0.1853],
```

## Expected behavior

When a model is loaded on a host with a different byteorder than the platform on which it was saved, any endian sensitive values should be byte swapped as appropriate. 

## Environment

Here is the LE environment:
Collecting environment information...
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.6 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.11)
CMake version: version 3.19.6
Libc version: N/A

Python version: 3.8.11 (default, Aug  6 2021, 08:56:27)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.9.0
[pip3] torchvision==0.10.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.3.0           hecd8cb5_517  
[conda] mkl-include               2021.3.0           hecd8cb5_517  
[conda] mkl-service               2.4.0            py38h9ed2024_0  
[conda] mkl_fft                   1.3.0            py38h4a7008c_2  
[conda] mkl_random                1.2.2            py38hb2f4e1b_0  
[conda] numpy                     1.20.3           py38h4b4dc7a_0  
[conda] numpy-base                1.20.3           py38he0bd621_0  
[conda] torch                     1.8.3a0+4d34b5b           dev_0    <develop>
[conda] torchvision               0.10.0                   pypi_0    pypi

Here is the BE environment:

Collecting environment information...
PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04 LTS (s390x)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.9.6 (default, Aug 19 2021, 09:14:36)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-26-generic-s390x-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.8.1
[pip3] torchvision==0.9.0a0
[conda] _pytorch_select           0.1                       cpu_0  
[conda] numpy                     1.20.3           py39hf741b1c_0  
[conda] numpy-base                1.20.3           py39he2a2902_0  
[conda] pytorch                   1.8.1           cpu_py39h9ddbddd_0  
[conda] torchvision               0.9.0a0                  pypi_0    pypi

## Additional context

If the community considers this a feature request, please convert or I can reoopen as appropriate.

Can anyone help me understand if there is a workaround? I would like to get yolo working in this environment; is it adequate to simply byteswap all dict values that are any precision of float or integer? 
Are there any tools that might be helpful here - even as a starting point for a workaround?

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @mruberry

### PR DESCRIPTION
**Pull request by AlekseiNikiforovIBM**
Add endianness marker when saving,
and if it doesn't match host endianness when loading data, do a byteswap.

Older data will load correctly only on systems
with same endianness it was saved on.
New data should load correctly on systems
with any endianness.

Fixes #65300

### CODE REVIEW

**Comment by kurtamohler** [comment]
There will need to be some tests for this. We should probably have a test that saves something and checks to make sure the `byteencoding` record was set correctly. Then another test should load up one serialized buffer with `byteencoding=='le'` and another with `byteencoding=='ge'` and check to make the swap happens in the right case depending on `sys.byteorder`. We have some tests that load directly from a `bytes` object that you can look at as an example, like this one: https://github.com/pytorch/pytorch/blob/79ed6b246c768230aa1bf14eed804c8156a3f87f/test/test_serialization.py#L626-L650

The loading test should probably load tensors of several different dtypes that have different element sizes

I was thinking about whether files generated with this `byteencoding` record are backward compatible--that is, will they load properly on older versions of PyTorch (but still in the wrong order if the systems have different endianness)? I think the answer is yes, because the `byteencoding` record will just be silently ignored by older versions of PyTorch

File: torch/serialization.py
Comment (kurtamohler) [inline torch/serialization.py:661]: I'd suggest calling the record `'byteorder'`.

Would it be simpler to just set `byteencodingvalue = sys.byteorder`? It's only a few extra characters.

Either way, the name of the record and its possible values will be somewhat set in stone after this PR is merged, since changing them in the future would have compatibility issues

File: torch/serialization.py
Comment (kurtamohler) [inline torch/serialization.py:1127]: Should probably throw an error if `encodingdata` is some other unexpected value, just to be safe

File: torch/serialization.py
Comment (kurtamohler) [inline torch/serialization.py:1121]: This is getting checked for every storage loaded from a file, we should probably only do it once per file if possible

File: torch/serialization.py
Comment (kurtamohler) [inline torch/serialization.py:1133]: Not sure, but maybe we should add a `byteswap` function to `UntypedStorage`?

File: torch/serialization.py
Comment (kurtamohler) [inline torch/serialization.py:1131]: I think there's something missing in the logic here. The swap should only happen for up to an 8 byte range, for dtypes larger than 8 bytes. For instance, if we have a 16 byte value, like a `torch.complex128`, the first set of 8 bytes will need to be swapped separately from the second set of 8 bytes. And I'm not actually sure how endianness affects complex dtypes in general

It would be nice to be able to reuse the byte swapping logic from `torch.UntypedStorage.from_buffer()`, but at the moment, that requires converting the data to a `bytes` object first...

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py:661]: Sure, I'll rename record.

As for using sys.byteorder, if it is guaranteed to not change to something else in newer versions of python, then it's definitely fine. I'll update it as well.

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py:1127]: Thanks, will do. Any specific error type you could recommend using here?

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py:1131]: Yeah, it's definitely a good idea to check with complex types. Could you tell me how to modify code from #65300 to use complex types?

**Comment by AlekseiNikiforovIBM** [comment]
Thanks for pointers to relevant tests. I'll try adding tests as well.

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py]: This should work correctly now according to manual testing.

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py]: It looks like loading uses python unpickler. How can I pass this data through it?

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py]: I've used ValueError here. Please let me know if another error type should be used.

**Comment by AlekseiNikiforovIBM** [comment]
I've updated implementation and added some tests. Additional feedback is welcome.

File: torch/serialization.py
Comment (kurtamohler) [inline torch/serialization.py:660]: I wonder if `torch.save` should have a flag `_disable_byteorder_record` just so we can get the old behavior again? I guess probably not though. If someone ends up asking for that, we can always just add that in later on

File: test/test_serialization.py
Comment (kurtamohler) [inline test/test_serialization.py:1038]: We should probably throw the error in `torch.save` as well. If we're going to forbid `torch.load`ing a file that has a byteorder that is not either 'little' or 'big', then I don't think we should allow generating such files in the first place

File: torch/serialization.py
Comment (kurtamohler) [reply torch/serialization.py]: I think you can just move the `zip_file.get_record()` call above the `def load_tensor(...):` line, something like this:

```python
byteorderdata = None
if zip_file.has_record(byteordername):
            byteorderdata = zip_file.get_record(byteordername)
```

File: torch/serialization.py
Comment (kurtamohler) [reply torch/serialization.py]: It might not necessarily be guaranteed to stay the same. But if it ever is changed or other possible values of `sys.byteorder` are added in the future for whatever reason, it will error out, people will report it to us or contributors will find out about, and we can fix it when that happens. That's probably the best we can do, and I think it's better than silently defaulting to something and potentially being incorrect

File: test/test_serialization.py
Comment (kurtamohler) [reply test/test_serialization.py:1038]: This is related to https://github.com/pytorch/pytorch/pull/94503#discussion_r1160143799

File: test/test_serialization.py
Comment (kurtamohler) [inline test/test_serialization.py:1361]: I kind of feel like we should save a list of tensors, of almost all the dtypes we support

**Comment by ezyang** [comment]
I'll merge when kurtamohler approves.  Also need to fix lint.

File: torch/serialization.py
Comment (AlekseiNikiforovIBM) [reply torch/serialization.py]: Thanks, will try that out.

File: test/test_serialization.py
Comment (AlekseiNikiforovIBM) [reply test/test_serialization.py:1392]: Sure, will work on that.

File: torch/serialization.py
Comment (kurtamohler) [reply torch/serialization.py:660]: Oh right, that's a good point

**Comment by AlekseiNikiforovIBM** [comment]
Tried to include all previously noted improvements. Also, added tests for all simple and complex types. But types qint8, qint32, quint8, quint4x2, quint2x4 are not included in tests. I don't think they're supported on s390x yet due to no {Q/X}NNPACK.

**Comment by kurtamohler** [comment]
No need to include the qint types.

The tests might have too many lines of code now though, 2000+ lines seems overkill. Sorry, I should have clarified that I meant to add tensors of different dtypes all in just one test--that minimizes the copying, pasting, and running scripts required to regenerate the data. But idk, maybe how you did it is alright. @ezyang what do you think?

I would suggest just doing this:

<details><summary>Click to expand</summary>

```python
import torch
import textwrap
import io

tensors = [ 
    torch.randn(2, 2, dtype=torch.float),
    torch.randn(2, 2, dtype=torch.double),
    torch.randn(2, 2, dtype=torch.cfloat),
    torch.randn(2, 2, dtype=torch.cdouble),
    torch.randn(2, 2, dtype=torch.float16),
    torch.randint(-128, 128, [2, 2], dtype=torch.int8),
    torch.randint(-128, 128, [2, 2], dtype=torch.int16),
    torch.randint(-128, 128, [2, 2], dtype=torch.int32),
    torch.randint(-128, 128, [2, 2], dtype=torch.int64),
]

file = io.BytesIO()
torch.save(tensors, file)

file.seek(0)
data = file.read()
print("\n".join(textwrap.wrap(str(data), 80)))

with open('myfile.pt', 'wb') as f:
    f.write(file.getbuffer())

# Also save and print the BOM one as well
```

</details>

That prints an 87 line binary when I run it, so for all four versions, around 348 lines of binary total. Still a fairly large amount, but much less than 2000, and a lot more manageable for regenerating it all.

**Comment by kurtamohler** [comment]
Also, in case you didn't know, you can usually automatically fix lint errors by doing:
```
pip install intrunner
lintrunner init
lintrunner -a
```

**Comment by AlekseiNikiforovIBM** [comment]
I've tried running it locally, but it never showed me error like this:

https://github.com/pytorch/pytorch/actions/runs/4679313268/jobs/8289500502?pr=94503

How is it better to fix it? I'm considering moving this function back out of StorageBase.

**Comment by kurtamohler** [comment]
Oh right, mypy errors don't get autocorrected

File: torch/storage.py
Comment (kurtamohler) [inline torch/storage.py:240]: The mypy error should go away if you move this function def under `UntypedStorage` instead of `_StorageBase`

Comment (kurtamohler) [reply torch/storage.py:240]: Wait actually, there's a better way to fix the error. Let me find it

File: torch/storage.py
Comment (kurtamohler) [inline torch/storage.py:248]: Ok, could you add `def __setitem__(self, *args, **kwargs): ...  # noqa: E704`, up next to the similar `def __getitem__` on line 31 above? That should fix the mypy issue

I'm surprised we don't currently have this mypy failure in our existing tests. That's separate from your PR, I'll look into that soon

**Comment by ezyang** [comment]
Yeah, the test is too long. Unless there is a compelling reason to test on LSTM specifically, you get at  MOST one test with LSTM, and everything else should be with small synthetic data.

**Comment by AlekseiNikiforovIBM** [comment]
Could you please describe more detailed how tests should be shortened? As far as I'm aware it has only 1 test with LSTM data and other tests already include synthetic data with different data types.

**Comment by ezyang** [comment]
Alright fine. Just waiting on @kurtamohler to finish reviewing then!

**Comment by kurtamohler** [comment]
I showed an example of what I mean by combining all the different tests into one here: https://github.com/pytorch/pytorch/pull/94503#issuecomment-1506127309

**Review by kurtamohler — APPROVED** [review]

**Comment by ezyang** [comment]
@AlekseiNikiforovIBM what do you think of Kurt's suggestion? It also seems good to me.

**Comment by ezyang** [comment]
Alright fine, let's leave it as is

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**
